### PR TITLE
feat(cohort): support event sequence parameter

### PR DESCRIPTION
**Pull request by EDsCODE**
## Problem

- use different base query when handling event sequence for optimal performance

## Changes

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

### CODE REVIEW

**Review by neilkakkar — COMMENTED** [review]
What are your thoughts on not going the funnel query route, and re-creating this here from scratch?

Just took a brief look, but it seems like integrating things into a single query would be much easier if we do it from scratch here instead, and reduce the number of subqueries we go through.

For example: Here's 2 queries, one using this PR (and then edits inline to make it a combination of other things as well), and the other using something built from scratch for this, taking advantage of the fact that we'll only have 2 events:

https://metabase.posthog.net/question/298-cohort-sequence-with-subquery

https://metabase.posthog.net/question/302-cohort-with-funnel-subquery-timing

(they're both similar time wise, and give the same result). Using the funnel builder implies going down to the `funnel_event_query` level, which is like 3 levels down (`ClickhouseFunnel` -> `base` -> `event_query` ), and then propagating it all back up, vs doing it in a single subquery.

File: ee/clickhouse/queries/cohort_query.py [hunk 1]
```diff
@@ -204,6 +209,7 @@ def _build_sources(self, subq: List[Tuple[str, str]]) -> Tuple[str, str]:
                 q += f"({subq_query}) {subq_alias}"
                 fields = f"{subq_alias}.person_id"
             elif prev_alias:  # can't join without a previous alias
+                # TODO: decide when to use inner join
```
[the remark this answers is not in the document]
Comment (EDsCODE) [reply ee/clickhouse/queries/cohort_query.py:212]: yup

**Comment by EDsCODE** [comment]
I would lean on the side of keeping it. I think keeping the logic reuse will be important in making sure the implementations across our system are consistent. This way if we decide to stir up the funnel query logic down the road, we will directly see the effects on cohorts (hopefully in tests and not in prod) and will make sure we keep our definition of "event sequence" consistent throughout the app

Also, if the results and performance are identical, from a practical standpoint I'd table the work for now so we can keep moving through the sprint as I don't think it's that heavy since `_get_sequence_query` is just reusing an interface?

**Comment by neilkakkar** [comment]
Sure, that works too. Just wanted to ensure we've considered both approaches. Looking forward to seeing how creating the rest of the cohort query via the funnel query looks like 👀 .

Another open question I had around re-using funnel query was that: do you think it will work easily with multiple sequences? Since we'd want to have all sequences inside the same subquery, and the funnel class seems too rigid (on a cursory-shallow glance) to support this bit too, since its not just extra fields, but extra events inside the class that need their own intervals and another earliest_timestamp date etc.

**Comment by EDsCODE** [comment]
The implementation in the PR as it is now should handle event sequence + any other condition actually. 

But, oof the multiple event sequences is the doozy for sure. This is the factor I keep overlooking 😢

**Comment by neilkakkar** [comment]
> should handle event sequence + any other condition actually.

Misread, it does indeed handle some cases, but its kinda awkward, since:

1. it misses out on events not in the funnel (say, a behavioural filter on an event other than a event inside the sequence wouldn't work)
2. Same for time ranges^, we'd want to earliest timestamp, like we were doing earlier

File: ee/clickhouse/queries/cohort_query.py [hunk 2]
```diff
@@ -204,7 +205,16 @@ def _build_sources(self, subq: List[Tuple[str, str]]) -> Tuple[str, str]:
...
+                    subq_alias == self.PERSON_TABLE_ALIAS
+                    and "person" not in [prop.type for prop in getattr(self._outer_property_groups, "flat", [])]
+                    and "cohort" not in [prop.type for prop in getattr(self._outer_property_groups, "flat", [])]
+                ):
+                    q = f"{q} {inner_join_query(subq_query, subq_alias, f'{subq_alias}.person_id', f'{prev_alias}.person_id')}"
+                else:
+                    q = f"{q} {full_outer_join_query(subq_query, subq_alias, f'{subq_alias}.person_id', f'{prev_alias}.person_id')}"
+
                 fields = if_condition(
                     f"{prev_alias}.person_id = '00000000-0000-0000-0000-000000000000'",
```
Comment (rcmarron) [inline ee/clickhouse/queries/cohort_query.py:219]: Unrelated to this PR, but why do we have this `person_id` check?

File: ee/clickhouse/queries/cohort_query.py [hunk 3]
```diff
@@ -474,52 +482,107 @@ def get_performed_event_first_time(self, prop: Property, prepend: str, idx: int)
...
+        step_cols.append(
+            f"if({seq_entity_query} AND timestamp > now() - INTERVAL {time_value} {time_interval}, 1, 0) AS {event_prepend}_step_1"
+        )
+        step_cols.append(f"if({event_prepend}_step_1 = 1, timestamp, null) AS {event_prepend}_latest_1")
+
+        return step_cols, condition_cols, aggregate_cols, {**entity_params, **seq_entity_params}
 
-        return query, params, self.FUNNEL_QUERY_ALIAS
+    def get_performed_event_sequence(self, prop: Property, prepend: str, idx: int) -> Tuple[str, Dict[str, Any]]:
+        return f"{self.SEQUENCE_FIELD_ALIAS}_{self.sequence_filters_lookup[str(prop.to_dict())]} = 2", {}
```
Comment (rcmarron) [inline ee/clickhouse/queries/cohort_query.py:585]: Super nit: it'd be nice to push the logic check for ` = 2` into the select part of the clause - just to keep the logic consistently up there.

**Review by rcmarron — COMMENTED** [review]
I just spent a bit with this PR, and it seems to work in all the cases I tested, but I can't say I've wrapped my head around our funnel query and what's happening in the nested queries + partition stuff, so I'd defer to @neilkakkar on the full review here.

File: ee/clickhouse/queries/cohort_query.py (hunk 2 above)
Comment (EDsCODE) [reply ee/clickhouse/queries/cohort_query.py:219]: When you do a full outer join and there's no match on the join condition, one of the fields will be null'd by default so you need to optimistically look for the id across the field from both tables

File: ee/clickhouse/queries/cohort_query.py (hunk 3 above)
Comment (EDsCODE) [reply ee/clickhouse/queries/cohort_query.py:585]: oh yes

**Comment by EDsCODE** [comment]
Addressed the feedback

File: ee/clickhouse/queries/funnels/base.py [hunk 4]
```diff
@@ -355,14 +355,15 @@ def _get_sorting_condition(self, curr_index: int, max_steps: int):
         return f"if({' AND '.join(conditions)}, {curr_index}, {self._get_sorting_condition(curr_index - 1, max_steps)})"
 
     def _get_inner_event_query(
-        self, entities=None, entity_name="events", skip_entity_filter=False, skip_step_filter=False
+        self, entities=None, entity_name="events", skip_entity_filter=False, skip_step_filter=False, extra_fields=[]
```
Comment (neilkakkar) [inline ee/clickhouse/queries/funnels/base.py:358]: looks like all of these changes aren't required anymore?

File: ee/clickhouse/queries/funnels/funnel.py [hunk 5]
```diff
@@ -186,15 +186,20 @@ def get_comparison_cols(self, level_index: int, max_steps: int):
 
         return ", ".join(cols)
 
-    def build_step_subquery(self, level_index: int, max_steps: int):
+    def build_step_subquery(
```
Comment (neilkakkar) [inline ee/clickhouse/queries/funnels/funnel.py:189]: same, touching any funnel class seems unnecessary?

**Review by neilkakkar — COMMENTED** [review]
Post humous, but LGTM, just 2 questions

**Pull request merged**